Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,143 @@
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(const HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& itemItr : firstTrackItems) {
+        auto* item = itemItr.key;
+        auto gridArea = itemItr.value;
+
+        insertIntoGridAndLayoutItem(item, gridArea);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        auto initialTrackSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(initialTrackSpan, itemSpan) : GridArea(itemSpan, initialTrackSpan);
+        insertIntoGridAndLayoutItem(item, area);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition)
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, const GridArea& area)
+{
+    m_renderGrid.currentGrid().insert(*child, area);
+
+    if (m_gridAxisDirection == ForColumns)
+        child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns));
+    else
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
+    child->layoutIfNeeded();
+    updateRunningPositions(child, area);
+}
+
+void GridMasonryLayout::updateRunningPositions(const RenderBox* child, const GridArea& area)
+{
+    auto gridSpan = m_masonryAxisDirection == ForRows ? area.columns : area.rows;
+
+    ASSERT(gridSpan.startLine() < m_runningPositions.size() && gridSpan.endLine() <= m_runningPositions.size());
+    gridSpan.clamp(m_runningPositions.size());
+
+    auto previousRunningPosition = m_runningPositions[gridSpan.startLine()];
+
+    auto newRunningPosition = m_masonryAxisDirection == ForRows ? previousRunningPosition + child->logicalHeight() + child->verticalMarginExtent() : previousRunningPosition + child->logicalWidth() + child->horizontalMarginExtent();
+    newRunningPosition += m_masonryAxisGridGap;
+
+    m_gridContentSize = std::max(m_gridContentSize, newRunningPosition - m_masonryAxisGridGap);
+
+    for (auto span : gridSpan)
+        m_runningPositions[span] = std::max(m_runningPositions[span], newRunningPosition);
+
+    updateItemOffset(child, previousRunningPosition);
```

Comment:
Should we ensure that we use 0 for items in the 'first track'?

It seems that we could have multiple overlapping items in the same cell (if they used definite positions), and I believe we still want to position all of these at the start edge, not successively?

Reply:
I believe that should already be happening. `m_runningPositions` should be initialized with 0 for each track before the algorithm runs and then `addItemsToFirstTrack` should just add those items to each appropriate grid axis track. As each first track item gets added, `m_runningPositions ` should then be updated for the track it was added to. Does that line up with what you were thinking?

Reply:
What if two items in the 'first track' are both in the same cell?

In the spec it has this: 'Note: This means that items with a definite placement at line 1 in the [masonry axis](https://drafts.csswg.org/css-grid-3/#masonry-axis) by [Grid Item Placement](https://drafts.csswg.org/css-grid-3/#grid-item-placement) step 1 can be made to intentionally overlap.'

Reply:
Do you have an example by chance that could recreate this scenario? For some reason I am not sure how this could occur

Reply:
```
<html><head><style>
grid {
  display: inline-grid;
  gap: 1px 2px;
  grid-template-columns: masonry;
  grid-template-rows: repeat(4,auto);
  border: 1px solid;
  padding: 0 1px 0 2px;
  vertical-align: top;
}

item {
  background-color: #444;
  color: blue;
  grid-row: 1;
  grid-column: 1;
}
</style>

</head><body><grid>
  <item>1</item>
  <item>2</item>
</grid></body></html>
```

Reply:
This was much more obvious than I thought... So if I am understanding the spec correctly, the running position should be computed to the max of those 2 items in that track after they have both been placed?

Reply:
That's my understanding, yes. And then non-first-track items should be positioned at the max of the running positions of all tracks they span.

Reply:
I tried to see if I could fix this really quickly but was running into some issues. I think I'm going to create a separate bug for this and work on it there